src/main/kotlin/com/openlattice/data/controllers/DataController.kt

```diff
@@ -0,0 +1,922 @@
+    }
+
+    override fun loadEntitySetData(
+            entitySetId: UUID, fileType: FileType?, token: String?
+    ): EntitySetData<FullQualifiedName> {
+        if (StringUtils.isNotBlank(token)) {
+            val authentication = authProvider
+                    .authenticate(PreAuthenticatedAuthenticationJsonWebToken.usingToken(token))
+            SecurityContextHolder.getContext().authentication = authentication
+        }
+        return loadEntitySetData(entitySetId, EntitySetSelection(Optional.empty()))
+    }
+
+    @Suppress("UNUSED")
+    @RequestMapping(
+            path = [ENTITY_SET + ENTITY_SET_ID_PATH],
+            method = [RequestMethod.POST],
+            consumes = [MediaType.APPLICATION_JSON_VALUE],
+            produces = [MediaType.APPLICATION_JSON_VALUE, CustomMediaType.TEXT_CSV_VALUE])
+    @Timed
+    fun loadEntitySetData(
+            @PathVariable(ENTITY_SET_ID) entitySetId: UUID,
+            @RequestBody(required = false) selection: EntitySetSelection,
+            @RequestParam(value = FILE_TYPE, required = false) fileType: FileType?,
+            response: HttpServletResponse
+    ): EntitySetData<FullQualifiedName> {
+        setContentDisposition(response, entitySetId.toString(), fileType)
+        setDownloadContentType(response, fileType)
+
+        return loadSelectedEntitySetData(entitySetId, selection, fileType)
+    }
+
+    override fun loadSelectedEntitySetData(
+            entitySetId: UUID, selection: EntitySetSelection, fileType: FileType?
+    ): EntitySetData<FullQualifiedName> {
+        return loadEntitySetData(entitySetId, selection)
+    }
+
+    private fun loadEntitySetData(entitySetId: UUID, selection: EntitySetSelection?): EntitySetData<FullQualifiedName> {
+        if (authz.checkIfHasPermissions(
+                        AclKey(entitySetId),
+                        Principals.getCurrentPrincipals(),
+                        EdmAuthorizationHelper.READ_PERMISSION
+                )) {
+            val entitySet = getEntitySet(entitySetId)
+
+            val entityKeyIds = if (selection == null) Optional.empty() else selection.entityKeyIds
+            val selectedProperties = getSelectedProperties(entitySetId, selection)
+
+            val normalEntitySetIds = if (entitySet.isLinking) entitySet.linkedEntitySets else setOf(entitySetId)
+
+            val authorizedPropertyTypesOfEntitySets = getAuthorizedPropertyTypesForEntitySetRead(
+                    entitySet, normalEntitySetIds, selectedProperties
+            )
+
+            val authorizedPropertyTypes = authorizedPropertyTypesOfEntitySets.values.first()
+            val orderedPropertyNames = LinkedHashSet<String>(authorizedPropertyTypes.size)
+
+            selectedProperties.filter { authorizedPropertyTypes.containsKey(it) }
+                    .map { authorizedPropertyTypes.getValue(it) }
+                    .map { pt -> pt.type.fullQualifiedNameAsString }
```

Comment:
I think you can combine this and the above `.map{}` ? Or will `.map{}` filter out nulls from `authorizedPropertyTypes.getValue(it)`?

Reply:
Well, getValue would throw an error if the key is not present in `authorizedPropertyTypes`, but generally speaking `.map{}` won't filter null values.
I will combine it into one map, but I think there shouldn't be much performance difference, since both do 2 computation c1 and c2 n times.

Reply:
Btw, I just ran a test for this with 20.000.000 size list and the combined mapping was slightly faster

Reply:
oh cool! always love to have some actual proof of these things